MatrixSDK/Data/RoomSummaryStore/CoreData/MXCoreDataRoomSummaryStore.swift

```diff
@@ -0,0 +1,280 @@
+    
+    private func deleteSummary(forRoomId roomId: String) {
+        let moc = defaultTempMoc
+        
+        moc.perform { [weak self] in
+            guard let self = self else { return }
+            if let existing = self.fetchSummary(forRoomId: roomId, in: moc) {
+                if let membersCount = existing.s_membersCount {
+                    moc.delete(membersCount)
+                }
+                if let trust = existing.s_trust {
+                    moc.delete(trust)
+                }
+                if let lastMessage = existing.s_lastMessage {
+                    moc.delete(lastMessage)
+                }
+                moc.delete(existing)
+            }
+            
+            self.saveIfNeeded(moc)
+        }
+    }
+    
+    private func deleteAllSummaries() {
+        let entityNames: [String] = [
+            MXRoomSummaryModel.entityName,
+            MXRoomLastMessageModel.entityName,
+            MXUsersTrustLevelSummaryModel.entityName,
+            MXRoomMembersCountModel.entityName
+        ]
+        
+        let moc = defaultTempMoc
+        
+        moc.perform { [weak self] in
+            guard let self = self else { return }
+            do {
+                for entityName in entityNames {
+                    let request = NSFetchRequest<NSFetchRequestResult>(entityName: entityName)
+                    let deleteRequest = NSBatchDeleteRequest(fetchRequest: request)
+                    try moc.execute(deleteRequest)
+                }
+                
+                self.saveIfNeeded(moc)
+            } catch {
+                MXLog.error("[MXCoreDataRoomSummaryStore] deleteAllSummaries failed: \(error)")
+            }
+        }
+    }
+    
+    private func allSummaries(_ completion: @escaping ([MXRoomSummaryProtocol]) -> Void) {
+        let request = MXRoomSummaryModel.typedFetchRequest()
+        
+        let moc = defaultTempMoc
+        
+        moc.perform {
+            do {
+                let results = try moc.fetch(request)
+                //  do not attempt to access other properties from the results
+                let mapped = results.compactMap({ MXRoomSummary(summaryModel: $0) })
+                DispatchQueue.main.async {
+                    completion(mapped)
+                }
+            } catch {
+                MXLog.error("[MXCoreDataRoomSummaryStore] fetchRoomIds failed: \(error)")
+            }
+        }
+    }
+    
+    private func saveIfNeeded(_ moc: NSManagedObjectContext) {
+        guard moc.hasChanges else {
+            return
+        }
+        moc.perform { [weak self] in
+            guard let self = self else { return }
+            do {
+                try moc.save()
+                //  propogate changes to the parent context, until reaching the persistent store
+                if let parent = moc.parent {
+                    self.saveIfNeeded(parent)
+                }
+            } catch {
+                moc.rollback()
```

Comment:
Yes, because the data flows into parent when saving, and vice versa when reading. Children are temporary, the source of truth is the most parent one, which is the persistent store. And also it should be very unlikely that a child is saved but parent failed because there is a merge policy between them.

Reply:
Ah got it! Thanks 🙂